content/blog/the-rise-of-special-project-infra.md

```diff
@@ -0,0 +1,168 @@
+
+Everybody needs a saga. Ours was: self-host a secure livestreaming server to accommodate a sizable audience participating from all around the world and at variable network speeds, _fast_. Oh, and add a chat room for synchronous Q&A.
+
+On the server side, we settled on the following stack: virtual private server (VPS) → NixOS → [Caddy](https://caddyserver.com/) → [Owncast](https://owncast.online/).
+
+The media would originate on our intrepid moderator’s local [OBS Studio](https://obsproject.com/) and be streamed at archival quality over the Real-Time Messaging Protocol (RTMP) to our remote Owncast service. Owncast would transcode the video to multiple stream qualities (i.e., optimized variants of video bitrate and CPU usage) and present the selected variant to the user in an attractive, branded web UI with a convenient chat box.
+
+### Phase II: Provision
+
+Controlling the narrative has always called for a bit of hardware. For this exercise, we went with a single VPS equipped with dedicated vCPUs. The general provisioning procedure follows:
+
+1. Choose a (preferably reputable and carbon negative) VPS provider.
+
+1. Provision an instance appropriate to the task and estimated audience.
+
+   - To minimize latency, select the available region that best approximates the location of the source stream as well as the mean location of the majority of expected participants where applicable. (Note: _object storage_ and _CDN caching_ additionally may be leveraged to accommodate participants; object storage enables the streaming service to serve media files when the concurrent user count exceeds available bandwidth, and CDN caching enables users to download these assets from the nearest available server.)
+   - Select a model with enough (preferably dedicated) vCPU cores to handle video transcoding, which is a highly processor-intensive operation. (Note: this should correspond to the stream variants selected in the [streaming service configuration](#streaming-service-configuration).)
+
+   The instance we provisioned for the first stress test included the following specs:
+
+   | vCPU |   RAM | Local storage | Location |
+   | ---: | ----: | ------------: | -------- |
+   |    8 | 32 GB |        240 GB | Germany  |
+
+1. Reserve a public IPv4 address and IPv6 subnet and associate with the provisioned instance.
+
+1. Assuming registration of the desired domain name, use a/the DNS registrar-provided interface to update the (or create an) A record for the target subdomain (`live.nixos.org` in our case) with the IP address associated with the provisioned instance. (Note that propagation may take up to 48 hours.)
+
+### Phase III: Configure
+
+Perhaps unintuitively, this was the easiest part. Thanks to the Nixpkgs community contributions of high-quality modules for great modern services, configuring a fully self-hosted livestreaming service safely behind a reverse proxy was scarcely more involved than toggling `enable`. Consider the following Nix code:
```

Comment:
```suggestion
Perhaps unintuitively, this was the easiest part. Thanks to the [`nixpkgs`](https://github.com/NixOS/nixpkgs) community contributions of high-quality modules for great modern services, configuring a fully self-hosted livestreaming service safely behind a reverse proxy was scarcely more involved than toggling `enable`. Consider the following Nix code:
```

Reply:
Would this additionally apply to “high-quality modules”, etc.? (Please see my reply to your comment on “greatness”.)

Reply:
While it's not clear what high quality would entail exactly (I think it would be valuable to readers if you could qualify this really briefly), "great" is just to broad and overloaded to really mean anything at all.

Again this is personal, and really not that important, but I am a bit annoyed every time someone uses "modern", because it also tells us nothing really, except that something is "recent", or currently en vogue for some reason. Being new or popular does not necessarily mean being good for any definition of "good" (except new or popular, that is).

That's why I suggest better not saying nothing at all. :)

Reply:
Thanks for the explanation. In this case, “great modern services” referred to the ones described in the article. Changing this to “many services” would’ve changed the meaning, so for the latest revision I took the long way around and added footnotes for clarity.